### ISSUE TITLE
cannot get the events sent to application insights

### ISSUE THREAD

**Issue by lehne** [serilog/serilog-sinks-applicationinsights#4]
Is there a way for me to call Flush() in order to send the events to Application Insights? Right now, I'm not seeing any show up even though I have the proper key set in the configuration.

**Comment by jbattermann** [comment]
Hi Jeff,

the AI sink is merely a very thin wrapper around the official Microsoft API and it sends the log messages your application writes to serilog to AI all non-exception events as [event telemetries](https://github.com/serilog/serilog-sinks-applicationinsights/blob/master/src/Serilog.Sinks.ApplicationInsights/Sinks/ApplicationInsights/ApplicationInsightsSink.cs#L86) (see [MS documentation](https://azure.microsoft.com/en-us/documentation/articles/app-insights-api-custom-events-metrics/)) and exception ones as [exception telemtries](https://github.com/serilog/serilog-sinks-applicationinsights/blob/master/src/Serilog.Sinks.ApplicationInsights/Sinks/ApplicationInsights/ApplicationInsightsSink.cs#L77).

I've just updated the sync to use the latest released & stable version of the AI client API but as this was just a bump fromn 1.2.0 > 1.2.3 I assume this won't change anything.

Could you check your AI portal / sub-section for your AI key whether the (custom) events do show up as events there?


Moreover, while the sink itself does not provide any buffering / flushing logic (unless you've wrapped it inside such a serilog sink), AI itself DOES buffer and send at times of its own choosing (see https://azure.microsoft.com/en-us/documentation/articles/app-insights-api-custom-events-metrics/#flushing-data) you might have to wait 'a little' and short-lived applications that end their process before the AI client API sent them over may not send at all - but again, that's MS implementation logic and nothing the sink does or can really influence (at the moment / as-is).

So basically the two important pieces of information for you might be: this sink sends the logged data as events or exceptions to AI and the AI client API used under the hood 'may' delay sending them over.

Let me know if this helped / you did find your events in the AI portal..

Cheers,
-Jörg

**Comment by lehne** [comment]
Hi Jorg,
I was able to get it to try to send the telemetry but the Application Insights endpoint is returning bad request. I think it is because the TimeStamp is not set properly. When I call Application Insights on my own it works because I can specify the TimeStamp.

**Comment by lehne** [comment]
And thank you!

**Comment by jbattermann** [comment]
Sure thing Jeff - and that's interesting.. digging a little deeper into that. However, I've been using the sink myself for a couple months now and do get events all just fine. Is there anything 'particular' about the logged data (i.e. particularly large messages, maybe some special encoding/character sets etc)?

**Comment by lehne** [comment]
My console outputs this for one of the events:
```C#
Application Insights Telemetry: {"name":"Microsoft.ApplicationInsights.Dev.myKey.Event","time":"0001-01-01T00:00:00.0000000+00:00","iKey":"myKey","data":{"baseType":"EventData","baseData":{"ver":2,"name":"Starting transfer processing task.","properties":{"LogLevel":"Information","LogMessage":"Starting transfer processing task.","LogTimeStamp":"01/09/2016 15:33:55 -05:00","DeveloperMode":"true"}}}}
```
Notice the time is 0. Not sure why it is doing this. No special encoding or large messages.

**Comment by jbattermann** [comment]
Yes that does seem odd - I'll fiddle with it and let you know what I've found / fixed.

**Comment by jbattermann** [comment]
Alright Jeff, I've updated the telemetry creation code to include a timestamp for both telemetry types and the package has been built & been published ([2.0.27](http://www.nuget.org/packages/Serilog.Sinks.ApplicationInsights/2.0.27)).. nuget.org might take a few minutes to include it in the index but could you give it a try then whether it fixes the problem now?

**Comment by lehne** [comment]
Works spectacularly well! Thank you Jörg!

**Comment by jbattermann** [comment]
Perfect, thanks for reporting it and glad I could help!
